_posts/2022-03-16-async-sink-base.md

```diff
@@ -0,0 +1,169 @@
+
+# Metrics
+
+There are three metrics that automatically exist when you implement sinks (and, thus, should not be implemented by yourself).
+
+* CurrentSendTime Gauge - returns the amount of time in milliseconds it took for the most recent request to write records to complete, whether successful or not.  
+* NumBytesOut Counter - counts the total number of bytes the sink has tried to write to the destination, using the method `getSizeInBytes` to determine the size of each record. This will double count failures that may need to be retried. 
+* NumRecordsOut Counter - similar to above, this counts the total number of records the sink has tried to write to the destination. This will double count failures that may need to be retried.
+
+# Sink Behavior
+
+There are six sink configuration settings that control the buffering, flushing, and retry behavior of the sink.
+
+* `int maxBatchSize` - maximum number of elements that may be passed in the list to submitRequestEntries to be written downstream.
+* `int maxInFlightRequests` - maximum number of uncompleted calls to submitRequestEntries that the SinkWriter will allow at any given point. Once this point has reached, writes and callbacks to add elements to the buffer may block until one or more requests to submitRequestEntries completes.
+* `int maxBufferedRequests` - maximum buffer length. Callbacks to add elements to the buffer and calls to write will block if this length has been reached and will only unblock if elements from the buffer have been removed for flushing.
+* `long maxBatchSizeInBytes` - a flush will be attempted if the most recent call to write introduces an element to the buffer such that the total size of the buffer is greater than or equal to this threshold value.
+* `long maxTimeInBufferMS` - maximum amount of time an element may remain in the buffer. In most cases elements are flushed as a result of the batch size (in bytes or number) being reached or during a snapshot. However, there are scenarios where an element may remain in the buffer forever or a long period of time. To mitigate this, a timer is constantly active in the buffer such that: while the buffer is not empty, it will flush every maxTimeInBufferMS milliseconds.
+* `long maxRecordSizeInBytes` - maximum size in bytes allowed for a single record, as determined by `getSizeInBytes()`.
+
+Destinations typically have a defined throughput limit and will begin throttling or rejecting requests once near. With multiple subtasks, we employ [Additive Increase Multiplicative Decrease (AIMD)](https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease) as a strategy for selecting the optimal batch size.
```

Comment:
I'm wondering what the meaning of "With multiple subtasks" is. Is this not used when the parallelism is 1?

Reply:
Sure, I will reword. It's just that if there were only 1 subtask, AIMD would not be needed for that subtask to determine the optimal throughput at which to write. However, with many subtasks with varying throughputs, AIMD is a strategy where, with no inter-subtask communication, an equilibrium write throughput is guaranteed to be reached for all subtasks.

Reply:
Btw, to clarify, when I say AIMD would not be needed when there is 1 subtask, I mean that particular strategy would not be needed. It's always advisable to have some form of rate limiting mechanism in play.